### ISSUE TITLE
[BUG] - When using Table, "type.getCollectionNode is not a function" error

### ISSUE THREAD

**Issue by Clovel** [nextui-org/nextui#729]
### Describe the bug

When using the `Table` components, I get an error in the console from inside NextUI's code : 

![image](https://user-images.githubusercontent.com/5107774/188172422-d53bc852-3afd-4254-8e8d-3c85787603c7.png)

I tried using the `Table` component two different ways : 

- First, by using the render function as indicated by the documentation :
```tsx
<Table.Header columns={counterTableColumns}>
  {
    (pColumn: CounterTableColumn) => {
      return (
        <Table.Column
          key={pColumn.uid}
          hideHeader={pColumn.hideHeader}
          align={pColumn.align ?? 'start'}
        >
          {pColumn.name}
        </Table.Column>
      );
    }
  }
</Table.Header>
```

- Then by using the `Array.map` method :

```tsx
<Table.Header>
  {
    counterTableColumns.map(
      (pColumn: CounterTableColumn) => {
        return (
          <Table.Column
            key={pColumn.uid}
            hideHeader={pColumn.hideHeader}
            align={pColumn.align ?? 'start'}
          >
            {pColumn.name}
          </Table.Column>
        );
      },
    )
  }
</Table.Header>
```

In both cases, no ESLint or TypeScript errors are displayed in my IDE nor in my terminal. 

These test were also made with the `Table.Body` section of the table.

I am using :
- Windows 10
- `"@nextui-org/react": "^1.0.0-beta.10"`
- `"react": "^18.2.0"`
- `"react-scripts": "5.0.1"`
- `"typescript": "4.4.2"`

### Steps to Reproduce the Bug or Issue

`CountersList.tsx` :

```tsx
/* Type declarations ----------------------------------- */
interface CounterTableColumn {
  name: string;
  uid: string;
  hideHeader?: boolean;
  align?: 'center' | 'start' | 'end';
}

/* Internal variables & constants ---------------------- */
const counterTableColumns: CounterTableColumn[] = [
  { name: 'Nom', uid: 'name' },
  { name: 'Description', uid: 'role' },
  { name: 'Actions', uid: 'actions', hideHeader: true },
];

const userCounters: CounterLightFragment[] = [
  {
    __typename: 'Counter',
    id: '1',
    name: 'Test1',
    description: 'Une description',
    avatarHash: null,
    avatarContentType: null,
  },
];

/* CountersList component prop types ------------------- */
interface CountersListProps {}

/* CountersList component ------------------------------ */
const CountersList: React.FC<CountersListProps> = () => {
  const userID = useReactiveVar(userIDReactiveVar);

  // const {
  //   data: userData,
  //   loading: loadingUserData,
  //   error: errorUserData,
  // } = useGetUserQuery(
  //   {
  //     variables: {
  //       // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  //       id: userID!,
  //     },
  //     skip: userID === null,
  //   },
  // );

  // const userCounters: CounterLightFragment[] = userData?.getUser?.counters?.map(
  //   (pVal: UserCounterOwnershipFragment) => {
  //     return pVal.counter;
  //   },
  // ) ?? [];

  console.log(`[DEBUG] <CountersList> userCounters :`, userCounters);

  return (
    <Table bordered>
      <Table.Header>
        {
          counterTableColumns.map(
            (pColumn: CounterTableColumn) => {
              return (
                <Table.Column
                  key={pColumn.uid}
                  hideHeader={pColumn.hideHeader}
                  align={pColumn.align ?? 'start'}
                >
                  {pColumn.name}
                </Table.Column>
              );
            },
          )
        }
      </Table.Header>
      {/* <Table.Header columns={counterTableColumns}>
        {
          (pColumn: CounterTableColumn) => {
            return (
              <Table.Column
                key={pColumn.uid}
                hideHeader={pColumn.hideHeader}
                align={pColumn.align ?? 'start'}
              >
                {pColumn.name}
              </Table.Column>
            );
          }
        }
      </Table.Header> */}
      <Table.Body>
        {
          userCounters.map(
            (pCounter: CounterLightFragment) => {
              return (
                <CountersListItem
                  key={`${pCounter.id}-${pCounter.name}`}
                  counter={pCounter}
                />
              );
            },
          )
        }
      </Table.Body>
      {/* <Table.Body items={userCounters}>
        {
          (pCounter: CounterLightFragment) => {
            return (
              <CountersListItem
                key={`${pCounter.id}-${pCounter.name}`}
                counter={pCounter}
              />
            );
          }
        }
      </Table.Body> */}
      <Table.Pagination
        shadow
        noMargin
        align="center"
        rowsPerPage={10}
        onPageChange={(page) => console.log({ page })}
      />
    </Table>
  );
};

/* Export CountersList component ----------------------- */
export default CountersList;
```

`CountersListItem.tsx` :

```tsx
/* CountersListItem component prop types --------------- */
interface CountersListItemProps {
  counter: CounterLightFragment;
}

/* CountersListItem component -------------------------- */
const CountersListItem: React.FC<CountersListItemProps> = ({ counter }) => {
  return (
    <Table.Row key={counter.id}>
      <Table.Cell>
        <User
          src={counter.avatarHash ?? undefined}
          name={counter.name}
        >
          {counter.description}
        </User>
      </Table.Cell>
      <Table.Cell>
        {JSON.stringify(counter)}
      </Table.Cell>
      <Table.Cell>
        Actions
      </Table.Cell>
    </Table.Row>
  );
};

/* Export CountersListItem component ------------------- */
export default CountersListItem;
```

### Expected behavior

Render the table correctly

### Screenshots or Videos

![image](https://user-images.githubusercontent.com/5107774/188174639-8524e995-4e07-42a0-a102-f61b5ecd19ad.png)

### Operating System Version

Windows 10

### Browser

Chrome

**Comment by jrgarciadev** [comment]
Hey @Clovel could you try again using `react:17`?, React 18 is still not fully supported by [react-aria](https://react-spectrum.adobe.com/react-aria/useTable.html) which is the library used by NextUI under the hood for creating tables, could you create a codesanbox/repository with the issue?

**Comment by Clovel** [comment]
Using React & React DOM 17 completely broke my application.

Most errors come from ReactNode types being incompatible with React.ReactNode types, etc. Most likely I have a bunch of incompatible dependencies now.

for example, I have MUI v5 in my project alongside NextUI. MUI v5.10 required a version of `@types/react-transition-group` that requires React 18.

```log
Type 'React.ReactNode' is not assignable to type 'import("C:/Users/<myUser>/repository/<project>/<project>-frontend/node_modules/@types/react-transition-group/node_modules/@types/react/index").ReactNode'.
```

Also, it seems that React 18 is listed as a peer dependency for the `react-aria` project you linked : 

```json
  "peerDependencies": {
    "react": "^16.8.0 || ^17.0.0-rc.1 || ^18.0.0",
    "react-dom": "^16.8.0 || ^17.0.0-rc.1 || ^18.0.0"
  },
```

I tried reproducing the error in a sandbox, but it's not the same error... Anyway : https://codesandbox.io/s/nextui-react-18-error-tes3wg

**Comment by Clovel** [comment]
So, I found that my issues where coming from a conflict of `@types/react` versions used by various dependencies. I had 3-4 different versions installed causing "Cant use as JSX component" or "React.ReactNode isn't compatible with "import(<dependencyPath>/@types/react/index).ReactNode" errors.

Now I finally have the same error as the CodeSandbox : 

![image](https://user-images.githubusercontent.com/5107774/189339457-dc7f9186-b63e-42f1-9ae7-b48e2d94c587.png)

I'll be creating a new issue for this. 

Meanwhile, I believe that this issue shouldn't be closed until `react-aria` is fully compatible with React 18, as said by @jrgarciadev

**Comment by reidbarber** [comment]
As mentioned in https://github.com/adobe/react-spectrum/issues/3502#issuecomment-1242241141, this error is due to using `CountersListItem` as children of `Table.Body`, when it's expecting `Table.Row` as children.

**Comment by Clovel** [comment]
> As mentioned in [adobe/react-spectrum#3502 (comment)](https://github.com/adobe/react-spectrum/issues/3502#issuecomment-1242241141), this error is due to using `CountersListItem` as children of `Table.Body`, when it's expecting `Table.Row` as children.

Even if the `CountersListItem` has `Table.Row` as it's main component ?

**Comment by reidbarber** [comment]
> Even if the `CountersListItem` has `Table.Row` as it's main component ? 

Correct. Refactoring around that should work

**Comment by jrgarciadev** [comment]
Hey @Clovel as @devongovett (react-aria lead maintainer) mentioned [here](https://github.com/adobe/react-spectrum/issues/3502), The `Table.Body` component only supports `Table.Row` elements as children

**Comment by TylerNRobertson** [comment]
Any reason for this limitation? I don't see a reason why we couldn't just pass the TableRow a custom `as` prop or something like that. Main reason for this is you cannot integrate NextUI tables with any other library that has container elements. For example: `react-beautiful-dnd` basically works out of the box with every UI library I've tried before aside from NextUI

Also curious how the builder works, if it looks for a `tr` element within the table body and then find the cells in that, what does allowing container elements do that's so bad? I think all that would need to happen in that case is changing how the builder selects the elements, instead of looking for direct child elements, select all the `tr` elements only for example

**Comment by defguru** [comment]
This behavior is non sense.

We can not do something like this, which would make sense if you need to do some processing per row, or event if you just want to encapsulate the row logic...

`
 const MyRow = ({item, ...tableRowProps} : {item: FooItemType}) => {
    const status = useFetchItemStatus(item)
    return (
      <TableRow {...tableRowProps} key={`example-${item.id}`}>
          <TableCell>{item.foo}</TableCell>
          <TableCell>{status || "loading satus"}</TableCell>
      </TableRow>
    )
 }

.... 
// Then in the parent component you would do
...

<Table aria-label="Example table">
    <TableHeader>
        <TableColumn>Col A</TableColumn>
        <TableColumn>Col B</TableColumn>
    </TableHeader>
    <TableBody>
        {data.map(item => (
            <MyRow item={item} />
        ))}
    </TableBody>
</Table>`

So honestly I don't understand the logic of not allowing this, this issue should remain opened imo since no solution was provided.

**Comment by devongovett** [comment]
It's not a limitation we wanted, just a technical trade off required by internal implementation details. React actually makes it extremely difficult to build components that are aware of their children like Table needs to be in order to implement things like keyboard navigation.

That said, in [React Aria Components](https://react-spectrum.adobe.com/react-aria/react-aria-components.html), our new component library built on top of React Aria hooks, we found a new way of implementing collection components that lifts this restriction. The goal is to extract this new implementation out once it is stable and offer it to React Aria hooks users such as Next UI as well. So hopefully in the near future Next UI will be able to lift the restriction as well.

In the meantime, we've recommended moving the logic to a component _inside_ the item rather than a wrapper. You could refactor your example like the following:

```jsx
<TableBody>
  {data.map(item => (
    <TableRow key={`example-${item.id}`}>
      <TableCell>{item.foo}</TableCell>
      <TableCell><ItemStatus item={item} /></TableCell>
    </TableRow>
  ))}
</TableBody>

function ItemStatus({item}) {
  const status = useFetchItemStatus(item)
  return status || 'loading status';
}
```

**Comment by dailytravel** [comment]
How I can add TableBody to <ReactSortable
                  list={(column?.tasks || []).map((task) => ({
                    ...task,
                    chosen: true,
                  }))}
                  setList={(newTasks) => {
                    setColumns((prevColumns) =>
                      prevColumns.map((prevColumn) =>
                        prevColumn.id === column.id
                          ? { ...prevColumn, tasks: newTasks }
                          : prevColumn
                      )
                    );
                  }}
                  group="tasks"
                  animation={150}
                  tag={TableBody}
                >

**Comment by spot-developers** [comment]
Have the same issue when trying to use DnD to reorder tabs. 
I'm assuming this is the same reliance on React Aria Components causing the underlying issue.
Look forward to the updates.

**Comment by sohrab-** [comment]
If the react-aria-components have found a way to support this, then why wouldn't NextUI?

This limitation results in more complex and less maintainable code, a scar on otherwise very clean and reusable component APIs.

**Comment by sohrab-** [comment]
What makes this quite cumbersome is that say if I have `AccordionItem`s that do not render due to some logic inside them, now all that logic needs to be pulled up into the parent component so the `AccordionItem`s can be conditionally removed. 😞

**Comment by rydonahue1** [comment]
So is there currently no way to breakout NextUI Table components into separate components? If so that is pretty disappointing considering the benefits you can gain from splitting things into server and client components.

**Comment by lazharichir** [comment]
Same issue with the `<Listbox>` component... Considering the entire point of React is to breakdown complex UI into smaller components, this is a bit of a weird and unexpected behaviour.

**Comment by divyam234** [comment]
@lazharichir  MenuItem also

**Comment by adityaa72** [comment]
This worked for me

```
import {
  SelectItemProps,
  SelectItem as SelectItemUI,
} from "@nextui-org/select";

type Props = SelectItemProps;

export const SelectItem = (props: Props) => {
  return <SelectItemUI {...props} />;
};
// @ts-expect-error Aria-Error
// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
SelectItem.getCollectionNode = SelectItemUI.getCollectionNode;

```

**Comment by christo9090** [comment]
Yeah tried using the table again and got this same error. Unfortunately, it makes this table component fairly useless. Especially if you need to manage state for individual rows. There are refactoring work arounds but most of them are anti-patterns for react. Love this library otherwise.

**Comment by Jaimayal** [comment]
Ran into this same issue, had to workaround using the nuqs library to re-render the table using on query parameters update. Hope we can get a proper way to implement logic/state per row soon!

**Comment by viinaji** [comment]
also got this issue with: DropdownItem

**Comment by alvaroaxo** [comment]
@viinaji I have the same problem with DropdownItem, were you able to solve it?

**Comment by viinaji** [comment]
@alvaroaxo My goal was to separate each item in the menu, However, I encountered that error when using the component. To address this, I turned to the collections feature [https://react-spectrum.adobe.com/react-stately/collections.html](url)
and the following is how I implemented it:
```TypeScript
<DropdownMenu aria-label="Card Menu" items={CardMenuSections()}>
  {section => (
    <DropdownSection {...section}>
      {/* @ts-ignore:disable-next-line */}
      {item => <DropdownItem {...item} />}
    </DropdownSection>
  )}
</DropdownMenu>
```
```TypeScript
type DropDownSectionType = {
  key: string;
  title?: string;
  showDivider?: boolean;
  items: (DropdownItemProps | undefined)[];
};
const CardMenuSections = (): DropDownSectionType[] => {
  return [
    {
      key: 'options',
      title: 'Options',
      showDivider: true,
      items: [LaunchConfig(), Extensions(), Pin()],
    },
    {
      key: 'update',
      title: 'Update',
      showDivider: true,
      items: [Update(), CheckForUpdate(), AutoUpdate()],
    },
    {
      key: 'about',
      title: 'About',
      items: [DocPage(), Info()],
    },
    {
      key: 'danger-zone',
      items: [Uninstall()],
    },
  ];
};
```
```TypeScript
export const Pin = (): DropdownItemProps => {
  const {cardId} = useCardData();

  const isPinned = useIsPinnedCard(cardId);

  const onPress = () => window.api.storageMgr.pinnedCards(isPinned ? 'remove' : 'add', cardId);

  return {
    key: 'pin-unpin',
    title: isPinned ? 'Unpin' : 'Pin',
    startContent: GetIconByName('Pin', {
      className: `${isPinned ? 'rotate-45' : 'rotate-0'} transition duration-500`,
    }),
    className: 'cursor-default',
    onPress,
  };
};
```

**Comment by codemonkeylabs-de** [comment]
> Why is this discussion marked as closed?

Indeed... I just ran into this issue myself. Pretty disappointing... 

Reading the history of the bug, it seems at some point the answer was "some smartie pants created some super smart scheme in some library and didn't think this thru so now this is a limitation, tough luck.. here is a workaround and go on with your lives"

Workaround is that instead of doing something like (in my case it happened with an Accordion component, but same thing applies to Tables etc)

```jsx
<Accordion>
  <MyPrettyComponent>
    { item => <AccordionItem ..} ... >...</AccordionItem> } 
  </MyPrettyComponent>
</Accordion>
```

You refactor this to:

```jsx
<MyNowNotSoPrettyComponent>{
  items =>
    <Accordion>{
        items.map(item => 
          <AccordionItem ...} ..> .. </AccordionItem>)
   }</Accordion>
 </MyNowNotSoPrettyComponent>
```

Worked for me.. and if you can't refactor your component for whatever reason, you are f.. ried.

**Comment by niraj-khatiwada** [comment]
I've found this issue to persist in almost every NextUI components that utilizes list items like Dropdown, ListBox, Select, Tabs, etc. To make a consistent theme in my App, I need to export the base component and all of its child component like this(image). This way I can just change any props from here and it'll be applied globally to my project. But this issue has created scattered abstraction everywhere I've to use the child component directly from nextui package instead of my `@components`. Really hoping for some help if someone finds any solution.
![image](https://github.com/nextui-org/nextui/assets/51436631/71286741-4af0-4942-9121-9b623df82347)

**Comment by saevarb** [comment]
@devongovett 

> It's not a limitation we wanted, just a technical trade off required by internal implementation details. React actually makes it extremely difficult to build components that are aware of their children like Table needs to be in order to implement things like keyboard navigation.
> 
> That said, in [React Aria Components](https://react-spectrum.adobe.com/react-aria/react-aria-components.html), our new component library built on top of React Aria hooks, we found a new way of implementing collection components that lifts this restriction. The goal is to extract this new implementation out once it is stable and offer it to React Aria hooks users such as Next UI as well. So hopefully in the near future Next UI will be able to lift the restriction as well.
> 
> In the meantime, we've recommended moving the logic to a component _inside_ the item rather than a wrapper. You could refactor your example like the following:
> 
> ```js
> <TableBody>
>   {data.map(item => (
>     <TableRow key={`example-${item.id}`}>
>       <TableCell>{item.foo}</TableCell>
>       <TableCell><ItemStatus item={item} /></TableCell>
>     </TableRow>
>   ))}
> </TableBody>
> 
> function ItemStatus({item}) {
>   const status = useFetchItemStatus(item)
>   return status || 'loading status';
> }
> ```

Are there any updates on this?

Honestly, this argument really rings hollow. Everyone in the react ecosystem knows how to handle this. You are partially there; you just need to use inversion of control.

Instead of making it impossible to use standard and good practice for building one's react components, you just document the requirements of any child component(i.e. it must be forwardref, it must take this set of props and spread them on the top level element, etc) and use render props. 

I've run into this issue trying to use almost every piece of this library and it's really making me reconsider using it, which is too bad. It was a candidate for becoming our primary internal component library at work and in my own personal projects. Unfortunately, the way it works now means that any react developer in our company that tries to use this library in a very sensible and reasonable way will run into extremely opaque errors. I would basically need to onboard everyone onto this issue specifically.

**Comment by devongovett** [comment]
> Are there any updates on this?

Yes! In our last React Aria release, we introduced a new collection implementation that supports custom wrappers, available in the `@react-aria/collections` package. Here's an example of using it with React Aria hooks: https://stackblitz.com/edit/react-aria-collections?file=src%2FApp.tsx,package.json&terminal=dev

Next UI will need to do some work to adopt this in their implementation. One wrinkle is that there is a small breaking API change: you'll need to pass an `id` prop instead of/in addition to a `key` since we can no longer access React's key in the new implementation (React does not expose it inside component props). This means Next UI will either need to release a major version, or provide a way to opt-into the new API. I'm happy to work with the team to help with that if needed. Otherwise, after the internal changes in Next UI are made, it should be pretty transparent to users.

**Comment by alexgradinar** [comment]
@niraj-khatiwada were you able to find a solution?

**Comment by niraj-khatiwada** [comment]
> @niraj-khatiwada were you able to find a solution?

Nope, no solution.

**Comment by fwcd** [comment]
Ran into this while trying to use `react-intersection-observer` to build a lazily loaded table by wrapping `<TableRow>`s in `<InView>`. Would be really useful to have the ability to wrap rows!

**Comment by domhhv** [comment]
I would also love to see a solution to this issue and an ability to move TableRow/TableCell to its own component, so that the various fetch requests of each table item is encapsulated.

**Comment by HasanHaghniya** [comment]
any updates?

**Comment by pppdns** [comment]
> Next UI will need to do some work to adopt this in their implementation. [...] This means Next UI will either need to release a major version, or provide a way to opt-into the new API.

Any updates?

**Comment by Dinesh-Kuniyal** [comment]
Any updates?

**Comment by rs-duongnt1** [comment]
Any updates?

**Comment by KindaBrazy** [comment]
I found a way in dropdown for bypass error
Instead of using this way:
```tsx
<DropdownMenu variant="faded" aria-label="Card Menu">
  <DropdownSection key="options" showDivider>
    <MenuLaunchConfig />
  </DropdownSection>
</DropdownMenu>
```
i used this way:
```tsx
<DropdownMenu variant="faded" aria-label="Card Menu">
  <DropdownSection key="options" showDivider>
    {MenuLaunchConfig()}
  </DropdownSection>
</DropdownMenu>
```
I'm not sure if this is the best or most correct way, but i'm no longer getting the error, and haven't get any performance issues, re-rendering problems, or state update problems.

**Comment by mbUSC** [comment]
Just adding my voice to the choir. I ran into this while using `<Listbox>` and it's awful.
Idk how this is still open since 2022, with all the heat it's been getting from the user base?

**Comment by mohidmakhdoomi** [comment]
New year, new update?

**Comment by almartino** [comment]
> I found a way in dropdown for bypass error Instead of using this way:
> 
> <DropdownMenu variant="faded" aria-label="Card Menu">
>   <DropdownSection key="options" showDivider>
>     <MenuLaunchConfig />
>   </DropdownSection>
> </DropdownMenu>
> i used this way:
> 
> <DropdownMenu variant="faded" aria-label="Card Menu">
>   <DropdownSection key="options" showDivider>
>     {MenuLaunchConfig()}
>   </DropdownSection>
> </DropdownMenu>
> I'm not sure if this is the best or most correct way, but i'm no longer getting the error, and haven't got any performance issues, re-rendering problems, or state update problems.

This is correct. By using this approach, it is possible to render a TableRow from a function call.
```typescript
return (
    <Table
      isHeaderSticky
      removeWrapper
      classNames={{
        base: "max-h-[520px] overflow-scroll",
        table: "min-h-[420px]",
      }}
    >
      <TableHeader columns={COMPETITORS_COLUMNS}>
        {(column) => <TableColumn key={column.key}>{column.label}</TableColumn>}
      </TableHeader>
      <TableBody>
        {competitors &&
          getGroups(competitors)
            .map((items) => RenderRows({ items, cellRenderer })) // here just don't use <RenderRows items={items} cellRenderer={cellRenderer}/>
            .flat()
            .flat()}
      </TableBody>
    </Table>
  );

// example of render rows
const RenderRows = ({
  items,
  cellRenderer,
}: {
  items: CompetitorAttribute[];
  cellRenderer: CellRenderer;
}) =>
  items.map((item, idx, values) => {
    const mainRow = CompetitorTableRow({
      item,
      values,
      idx,
      cellRenderer,
    });

    if (values.length === 1 || idx > 0) {
      return [mainRow];
    }

    return [MoveHereRow(), mainRow];
  });
```

`MoveHereRow` and `mainRow` wraps <TableRow>
```typescript
// MoveHereRow example
const MoveHereRow = () => {
  return (
    <TableRow>
      <TableCell>
        <Button>Move here </Button>
      </TableCell>
      <TableCell> </TableCell>
      <TableCell> </TableCell>
      <TableCell> </TableCell>
      <TableCell> </TableCell>
      <TableCell> </TableCell>
      <TableCell> </TableCell>
    </TableRow>
  );
};
```

**Comment by lsaavedr** [comment]
> I found a way in dropdown for bypass error Instead of using this way:
> 
> <DropdownMenu variant="faded" aria-label="Card Menu">
>   <DropdownSection key="options" showDivider>
>     <MenuLaunchConfig />
>   </DropdownSection>
> </DropdownMenu>
> i used this way:
> 
> <DropdownMenu variant="faded" aria-label="Card Menu">
>   <DropdownSection key="options" showDivider>
>     {MenuLaunchConfig()}
>   </DropdownSection>
> </DropdownMenu>
> I'm not sure if this is the best or most correct way, but i'm no longer getting the error, and haven't got any performance issues, re-rendering problems, or state update problems.

Thank you very much, I believe this issue breaks many common practices in using React, and for future projects, I will need to evaluate another component library that doesn't depend on react-aria :(

**Comment by Vanortton** [comment]
> I found a way in dropdown for bypass error Instead of using this way:
> 
> <DropdownMenu variant="faded" aria-label="Card Menu">
>   <DropdownSection key="options" showDivider>
>     <MenuLaunchConfig />
>   </DropdownSection>
> </DropdownMenu>
> i used this way:
> 
> <DropdownMenu variant="faded" aria-label="Card Menu">
>   <DropdownSection key="options" showDivider>
>     {MenuLaunchConfig()}
>   </DropdownSection>
> </DropdownMenu>
> I'm not sure if this is the best or most correct way, but i'm no longer getting the error, and haven't got any performance issues, re-rendering problems, or state update problems.

Worked for me, thank you!

**Comment by everurstruly** [comment]
> I found a way in dropdown for bypass error Instead of using this way:
> 
> <DropdownMenu variant="faded" aria-label="Card Menu">
>   <DropdownSection key="options" showDivider>
>     <MenuLaunchConfig />
>   </DropdownSection>
> </DropdownMenu>
> i used this way:
> 
> <DropdownMenu variant="faded" aria-label="Card Menu">
>   <DropdownSection key="options" showDivider>
>     {MenuLaunchConfig()}
>   </DropdownSection>
> </DropdownMenu>
> I'm not sure if this is the best or most correct way, but i'm no longer getting the error, and haven't got any performance issues, re-rendering problems, or state update problems.

This worked well for static elements until I added event handlers. Never got fired

**Comment by luixo** [comment]
Seems like this was [updated](https://github.com/adobe/react-spectrum/issues/2746) for React Aria Components, can similar approach be used here?

**Comment by PerryFinn** [comment]
Any updates?

I am using the `@heroui/select` while using the `code-inspector-plugin` library, which has caused the same error to appear in my application, which is very frustrating. Before using the heroui component library, this error did not occur.

I really like the style of the heroui component library, it's full of modernity, and I hope this issue can be fixed as soon as possible.

**Comment by ATT273** [comment]
i got this error while trying to spit TableRow to separated component
```typescript
const Table = () => {
<Table aria-label="Role list">
   <TableHeader>
        {columns.map((column) => (
      <TableColumn key={column.key}>{column.label}</TableColumn>
    ))}
  </TableHeader>
  <TableBody emptyContent="No data">
    {roles && roles.length > 0 ? roles.map((item) => <RoleTableRow key={item.code} item={item} />) : []}
  </TableBody>
</Table>
}

// TableRow 
const RoleTableRow = () => {
  return (
    <TableRow>
      <TableCell className="text-center">My data</TableCell>
    </TableRow>
  );
};
``` 
Instead of using it as a component, use it as a function
```typescript
 <TableBody emptyContent="No data">
    {roles && roles.length > 0 ? roles.map((item) => RoleTableRow({ item })) : []}
  </TableBody>
```

**Comment by criskell** [comment]
> Seems like this was [updated](https://github.com/adobe/react-spectrum/issues/2746) for React Aria Components, can similar approach be used here?

It looks like they haven't updated the hooks API to use the new collections API.

**Comment by criskell** [comment]
> > Are there any updates on this?
> 
> Yes! In our last React Aria release, we introduced a new collection implementation that supports custom wrappers, available in the `@react-aria/collections` package. Here's an example of using it with React Aria hooks: https://stackblitz.com/edit/react-aria-collections?file=src%2FApp.tsx,package.json&terminal=dev
> 
> Next UI will need to do some work to adopt this in their implementation. One wrinkle is that there is a small breaking API change: you'll need to pass an `id` prop instead of/in addition to a `key` since we can no longer access React's key in the new implementation (React does not expose it inside component props). This means Next UI will either need to release a major version, or provide a way to opt-into the new API. I'm happy to work with the team to help with that if needed. Otherwise, after the internal changes in Next UI are made, it should be pretty transparent to users.

Is there a way to do this, HeroUI team? Any updates?

Please 🙏 

---

Edit:

It looks like in v3 they're going to use RAC with the new collections API for Table components, I think this would work if they used it:
https://react-spectrum.adobe.com/react-aria/Table.html#reusable-wrappers

v3 branch:
https://github.com/heroui-inc/heroui/tree/v3/packages/react

**Comment by Carlglain** [comment]
Any updates?
Cause i tried enforcing my Casl permissions on the Tab components like this

<img width="952" height="640" alt="Image" src="https://github.com/user-attachments/assets/2bad2ef7-b37c-4fa9-ac8e-89c812153092" />

and i still got thesame error

TypeError: type.getCollectionNode is not a function
    at $eb2240fc39a57fa5$export$bf788dd355e3a401.getFullNode (
)
    at getFullNode.next (<anonymous>)
    at $eb2240fc39a57fa5$export$bf788dd355e3a401.iterateCollection (
)
    at iterateCollection.next (<anonymous>)
    at [Symbol.iterator] (
)
    at [Symbol.iterator].next (<anonymous>)
    at new $a02d57049d202695$export$d085fb9e920b5ca7 (
)
    at

    at

    at mountMemo (
)
    at Object.useMemo (
)
    at exports.useMemo (
)
    at $7613b1592d41b092$export$6cd28814d92fa9c9 (
)
    at $e72dd72e1c76a225$export$2f645645f7bca764 (
)
    at $a0d645289fe9b86b$export$e7f05e985daf4b5f (
)
    at $76f919a04c5a7d14$export$4ba071daf4e486 (
)
    at useTabs (
)
    at Tabs2 (
)
    at Object.react_stack_bottom_frame (
)
    at renderWithHooks (
)
    at updateForwardRef (
)
    at beginWork (
)
    at runWithFiberInDEV (
)
    at performUnitOfWork (
)
    at workLoopSync (
)
    at renderRootSync (
)
    at performWorkOnRoot (
)
    at performSyncWorkOnRoot (
)
    at flushSyncWorkAcrossRoots_impl (
)
    at processRootScheduleInMicrotask (
)

**Comment by Carlglain** [comment]
> Any updates? Cause i tried enforcing my Casl permissions on the Tab components like this
> 
> <img alt="Image" width="952" height="640" src="https://private-user-images.githubusercontent.com/122307101/562700012-2bad2ef7-b37c-4fa9-ac8e-89c812153092.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzMzNzEwNTcsIm5iZiI6MTc3MzM3MDc1NywicGF0aCI6Ii8xMjIzMDcxMDEvNTYyNzAwMDEyLTJiYWQyZWY3LWIzN2MtNGZhOS1hYzhlLTg5YzgxMjE1MzA5Mi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwMzEzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDMxM1QwMjU5MTdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wYzc2ODkyMDM2YWJhZDllOWYzZTcwNjY3YzcyNDZhMzgzMzYyZGU2ZmNlMWQ2ZmJkYzgxMWUxYjZiMWI5NGY4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.Ay5rpXXgQaCFznOojM-5bA-JUG9XVNs6TdFnLqYEd50">
> and i still got thesame error
> 
> TypeError: type.getCollectionNode is not a function     at $eb2240fc39a57fa5$export$bf788dd355e3a401.getFullNode ( )     at getFullNode.next ()     at $eb2240fc39a57fa5$export$bf788dd355e3a401.iterateCollection ( )     at iterateCollection.next ()     at [Symbol.iterator] ( )     at [Symbol.iterator].next ()     at new $a02d57049d202695$export$d085fb9e920b5ca7 ( )     at
> 
>     at
> 
>     at mountMemo ( )     at Object.useMemo ( )     at exports.useMemo ( )     at $7613b1592d41b092$export$6cd28814d92fa9c9 ( )     at $e72dd72e1c76a225$export$2f645645f7bca764 ( )     at $a0d645289fe9b86b$export$e7f05e985daf4b5f ( )     at $76f919a04c5a7d14$export$4ba071daf4e486 ( )     at useTabs ( )     at Tabs2 ( )     at Object.react_stack_bottom_frame ( )     at renderWithHooks ( )     at updateForwardRef ( )     at beginWork ( )     at runWithFiberInDEV ( )     at performUnitOfWork ( )     at workLoopSync ( )     at renderRootSync ( )     at performWorkOnRoot ( )     at performSyncWorkOnRoot ( )     at flushSyncWorkAcrossRoots_impl ( )     at processRootScheduleInMicrotask ( )

i had to use a map to by pass this 

<img width="758" height="424" alt="Image" src="https://github.com/user-attachments/assets/5aa8a3d3-b863-4e73-8f05-a86f51133f91" />

**Comment by wingkwong** [comment]
Closing in favor of v3. If the issue is reproducible in v3, please create a new one with a minimal reproducible example.
